[REF/#182] Refactor navigation with AppState and simplify dialog handling

angryPodo opened the pull request:
### Related issue

   * closed #182
### Work Description ✏️

   * `MainNavigator`를 `MainAppState`로 통합 및 리팩토링
       * 기존 `MainNavigator`의 탐색 관련 책임과 UI 상태 로직을 `MainAppState`로
         이전하여 역할을 통합했습니다.
       * `MainAppState`를 UI 상태 홀더(State Holder)로 사용하여, UI에 필요한 모든
         상태(현재 탭, 하단 바 표시 여부, 네트워크 상태 등)를 중앙에서 관리하도록
         개선했습니다.
   * UI 상태 로직에서 `@Composable` 종속성 제거
       * `currentTab`, `isBottomBarVisible` 등을
         `navController.currentBackStackEntryFlow`를 `StateFlow`로 변환하여 제공하도록
         변경했습니다.
       * 이를 통해 상태 로직이 더 이상 Compose 런타임에 종속되지 않아 테스트
         용이성이 향상되고, 책임 분리가 명확해졌습니다.
   * `MainViewModel` 제거
       * 기존 MainViewModel에서 처리하던 네트워크 상태 관찰 로직을 MainAppState로
         이동시켜, 불필해진 ViewModel을 제거했습니다.
   * 다이얼로그 시스템 리팩토링
       * 기존의 DialogController, LocalDialogController를 제거하고, 상태 기반으로
         동작하는 DialogState와 DialogEventProvider를 도입했습니다.
       * 이제 각 화면에서는 LocalDialogEventProvider를 통해 다이얼로그 이벤트를
         발생시키기만 하면 되어, 구현이 더 간결해졌습니다.

 ### To Reviewers 💌
  `MainNavigator` → `MainAppState`로의 구조 변경 및 내비게이션 로직 정리에 중점을
  두고 봐주세요.

  현우형이 말한 '땜빵 코드'의 이유를 쭉 생각해봤는데, 아마 `MainNavigator`라는
  일반 클래스가 `@Composable` 함수에 직접 의존해서 책임이 불분명하고 테스트가 
  어려운 구조였기 때문이 아닐까 싶습니다

  그래서 이번에 '컴포즈답게' 관리하자는 생각으로 NiA의 State
  Holder 패턴을 참고해서 `MainAppState`로 싹 바꿨습니다. 이제 `MainAppState`가 UI
  상태를 책임지고, 데이터는 `Flow`를 통해 단방향으로 흐르게 되어 훨씬 깔끔하고
  예측 가능한 구조가 되었다고 생각합니다...

겸사겸사 다이얼로그도 변경했습니다ㅎㅎ 기존코드는 너무 결합이 강했어서..

angryPodo committed:
[REF/#182] Rename MainNavigator to MainAppState

This commit refactors `MainNavigator` to `MainAppState` and updates its usages.
The changes include:
- Renaming `MainNavigator.kt` to `MainAppState.kt`.
- Updating `MainAppState` to use `CoroutineScope` for managing state flows for `currentTab` and `isBottomBarVisible`.
- Simplifying navigation options logic in `MainAppState.navigate`.
- Removing `@Composable` annotation from `MainTab.find` and `MainTab.contains` as they no longer rely on composable context.

angryPodo committed:
[MOD/#182] Adjust NavOptions for Voca tab navigation

This commit modifies the navigation behavior for the Voca tab.
Previously, all tabs shared the same NavOptions. Now, the Voca
tab utilizes a distinct `vocaNavOptions` instance.

angryPodo committed:
[REF/#182] Move ViewModel logic to AppState and simplify navigation

- Deleted `MainViewModel` and transferred its network monitoring logic to `MainAppState`.
- `MainAppState` now directly observes network status.
- Simplified bottom tab navigation logic in `MainAppState` by using `popUpTo(navController.graph.findStartDestination().id)` for consistent behavior.
- Updated `MainScreen` and `MainActivity` to reflect these changes, passing `NetworkMonitor` to `MainAppState`.

angryPodo committed:
[REF/#182] Replace DialogController with DialogEventProvider

This commit refactors the dialog handling mechanism by replacing `DialogController` with `DialogEventProvider`.

The key changes are:
- `DialogController` and `LocalDialogController` have been removed.
- `DialogEventProvider`, `DialogState`, and `LocalDialogEventProvider` have been introduced.
- `DialogEventProvider` separates the dialog state (`DialogState`) from the actions to show/dismiss the dialog. This promotes a more unidirectional data flow.
- `MainAppState` now manages `DialogState` and provides functions to show/dismiss dialogs.
- `HilingualErrorDialog` now accepts `DialogState` and an `onDismiss` lambda instead of a `DialogController`.
- Screens that previously used `LocalDialogController` now use `LocalDialogEventProvider` to interact with dialogs.

This change aims to improve the clarity and maintainability of dialog management within the application.

angryPodo committed:
[MOD/#182] Improve MainTab navigation logic

angryPodo committed:
[MOD/#182] Improve MainTab navigation logic

nahy-512 commented on core/designsystem/src/main/java/com/hilingual/core/designsystem/event/DialogProvider.kt:
```diff
@@ -16,31 +16,26 @@
-        isVisible = true
-    }
+@Immutable
+data class DialogState(
+    val isVisible: Boolean = false,
+    val onClickAction: () -> Unit = {}
+)
```
p3
State는 data class로 만드는 게 일반적일까요?
저는 뭔가 '클릭 액션도 데이터인가?' 싶어서 조금 낯설게 느껴지기도 하는데, data class로 만드신 이유가 있는지 궁금합니다!

angryPodo replied:
해당 `DialogState`를 `data class`로 만든 이유는 `isVisible` 값을 기반으로 UI를 갱신해야 하고, 동시에 copy() 등을 활용해 쉽게 상태를 업데이트하거나 비교할 수 있도록 하기 위함입니다

nahy-512 commented on presentation/main/src/main/java/com/hilingual/presentation/main/MainAppState.kt:
```diff
@@ -16,38 +16,81 @@
-        @Composable get() = MainTab.find { tab ->
-            currentDestination?.hasRoute(tab::class) == true
+    val isOffline: StateFlow<Boolean> = networkMonitor.isOnline
+        .map(Boolean::not)
```
p3
`NetworkMonitor`에서 쓰는 `isOnline`을 오프라인 기준으로 바꿔주기 위해서 여기서 `map`을 사용하신 부분이 인상깊네요ㅎㅎ
초기에 NetworkMonitor 로직 짜실 때 NetworkMonitor에서 `isOffline`이 아니라 `isOnline`을 반환하게 한 이유가 있을까요? 네트워크 상태를 MainAppState 여기서 전역적으로 관리해주고 있는 걸로 보여 한번만 쓰이는 로직일 것 같다만, 그래도 뭔가 저희 앱을 사용하기 위해서는 온라인 상태여야 하니까 `isOffline`을 예외 상황으로 보고 처리해주고 있지 않나 싶어서요!

nahy-512 replied:
아니면 추후 단어장을 오프라인 상태에서도 볼 수 있게 해야돼서 룸디비를 쓴다고 하면..
isOnline 여부에 따라 '서버 API 호출' or '룸디비에서 데이터 가져옴' 이렇게도 나눌 수 있을랑가요

angryPodo replied:
말씀해주신 대로 `NetworkMonitor`에서 `isOffline`을 바로 제공하는 방법도 충분히 고려할 수 있었어요.
그럼에도 `isOnline`을 기본으로 제공하고 `MainAppState`에서 `map`을 통해 변환하도록 설계한 데에는 **책임 분리 원칙을 명확히 하고 싶었던 의도**가 있었어요.
### 1. `NetworkMonitor`의 책임: 객관적인 사실 전달
`NetworkMonitor`의 역할은 이름 그대로 **네트워크 상태를 객관적으로 감시**해요.
이 클래스가 제공해야 할 가장 근본적이고 중립적인 정보는 **"네트워크에 연결되어 있는가?"** 라는 질문에 대한 답, 즉 `isOnline`이었어요.

* **긍정 상태 우선**:
  기본적으로 `isOnline`이라는 긍정 상태를 제공하는 것이 더 직관적이고 재사용성이 높다고 생각했어요. 예를 들어, 나중에 *"온라인일 때만 데이터를 동기화"* 하는 기능이 생긴다면 `isOnline` 상태를 그대로 사용할 수 있었을 거예요. `NetworkMonitor`가 `isOffline`만 제공했다면 `!isOffline`으로 한 번 더 변환해야 했을 거예요.

* **중립성**:
  `NetworkMonitor`는 이 정보가 UI에서 '오프라인 다이얼로그'를 띄우는 데 쓰일지, 아니면 다른 기능에 쓰일지 몰라야 한다고 생각해요.
  그저 **순수한 사실만 제공하는 것**이 이 클래스의 명확한 책임이겠죠?

### 2. `MainAppState`의 책임: UI를 위한 상태 가공

`MainAppState`의 역할은 다양한 데이터 소스(여기서는 `NetworkMonitor`)로부터 받은 정보를 **UI가 소비하기 좋은 상태로 가공하는 것**이었어요.
* **UI의 관심사**:
  우리 UI의 현재 관심사는 \*"사용자가 오프라인일 때, 에러 다이얼로그를 보여줘야 한다"\*였어요. 즉, UI는 `isOffline`이라는 상태에 직접적으로 반응해야 했어요. 따라서 `'온라인인가?'(isOnline)`라는 객관적인 사실을 `'UI가 오프라인 상태에 대처해야 하는가?'(isOffline)`라는 **UI 특화 상태**로 변환하는 작업은, UI 상태 홀더인 `MainAppState`가 수행하는 것이 가장 적절하다고 판단했어요.

angryPodo replied:
> 아니면 추후 단어장을 오프라인 상태에서도 볼 수 있게 해야돼서 룸디비를 쓴다고 하면..
> isOnline 여부에 따라 '서버 API 호출' or '룸디비에서 데이터 가져옴' 이렇게도 나눌 수 있을랑가요

이 부분은 추후에 개발해야겠지만...예외케이스로 나눠야 할거같네요. 네트워크 요청시에 오프라인 일경우만 다이얼로그를 띄운다던지

nahy-512 replied:
우와.. 이런 사소한 부분을 코멘트 달아도 될까 고민하다가 변수 선언 부분에서도 민재님의 어떤 의도가 담겨있을까 궁금해서 남겨본 거였는데ㅎㅎ 제 생각보다도 훨씬 많은 의도가 담겨 있었네요!!  적어주신 긍정 상태 우선과 관심사 분리가 그간 개발하면서 많이 접했던 내용인데 이런 식으로 하이링구얼에서 또 활용되었군요. 길게 답변 남겨주셔서 감사합니다☺️ 덕분에 명확히 이해됐어요

nahy-512 commented on presentation/main/src/main/java/com/hilingual/presentation/main/MainAppState.kt:
```diff
@@ -16,38 +16,81 @@
+        .map(Boolean::not)
+        .stateIn(
+            scope = coroutineScope,
+            started = SharingStarted.WhileSubscribed(5_000),
```
p3
숫자에 이렇게 언더스코어 넣는 걸 선호하시나요??

angryPodo replied:
저는 별론데 넣더라구요?

Hyobeen-Park commented on presentation/main/src/main/java/com/hilingual/presentation/main/MainScreen.kt:
```diff
@@ -67,17 +67,22 @@ private const val EXIT_MILLIS = 3000L
-    navigator: MainNavigator = rememberMainNavigator(),
-    viewModel: MainViewModel = hiltViewModel()
+    networkMonitor: NetworkMonitor,
+    appState: MainAppState = rememberMainAppState(networkMonitor = networkMonitor)
```
P1: networkMonitor를 rememberMainAppState(networkMonitor = networkMonitor) 여기에서만 사용하는 것 같은데 그렇다면 state를 activity에서 받아오는 것이 아닌 MainScreen 내부에서 기본값으로 설정하신 이유가 있으신가요? 내부에서 사용하는 값이 아닌데 스크린이 받아오는 것이 조금 마음에 걸려서요..!!

Hyobeen-Park replied:
```suggestion
    appState: MainAppState
```

이렇게만 받아오고 MainActivity에서 state를 넣어주는 구조도 고려해보셨을까요ㅕ?? 테스트 코드를 추가할지는 모르겠지만 이런 구조가 조금 더 테스트 하기 용이할 것 같아서요! networkMonitor를 Screen 내부에서 사용하는 것이 아니라면 이런 방식을 고려해보는 것도 좋을 것 같아서 질문 드립니다!

angryPodo replied:
좋은 의견 감사합니다ㅎㅎ 그러게요...이렇게 안바꿀 이유가 없네요ㅎㅎ 상태 호이스팅해놓겠습니다ㅎㅎ

Hyobeen-Park commented on presentation/main/src/main/java/com/hilingual/presentation/main/MainScreen.kt:
```diff
@@ -182,7 +187,10 @@ internal fun MainScreen(
+            HilingualErrorDialog(
+                state = appState.dialogState,
+                onDismiss = appState::dismissDialog
+            )
```
P2: 구조를 제가 잘 이해한건지 모르겠는데요 LocalDialogEventProvider로 provider 받아서 show하면 이 에러 다이얼로그만 뜨는거 맞나요..?? 다른 다이얼로그를 사용하려면 provider 없이 사용하거나 새로운 provider 객체를 만들어서 뿌려줘야 하는 것 같아서요 제가 제대로 이해한게 맞나요...??

angryPodo replied:
네 맞아요 지금 에러 다이얼로그가 하나로 통일 되어있어서요. 스낵바 생각해보시면 텍스트만 바뀌듯이 람다만 바뀌는 구조입니다. 이건 기획측에서 바꿀때 바뀌지 않을까요? 지금은 에러케이스는 하나로 통일이니까...

Hyobeen-Park reviewed (commented):
코드가 재밌어요!!ㅋㅋㅋㅋ

nahy-512 commented on presentation/main/src/main/java/com/hilingual/presentation/main/MainAppState.kt:
생각하지 못했던 코드 형태인데 읽어보면서 엄청 신기했네요. 특히 네비게이터에 `@Composable` 종속성 다 없애주신 부분들!
언급해주신 거 보고 NiA 코드를 좀 살펴봤는데요. NiA에서도 `currentDestination` 구하는 부분이나.. 이런 부분에서 컴포저블 종속성이 좀 있더라구요! 근데 이번 민재님 작업에서는 안 보이는 거 같아서 추가로 참고하신 다른 자료가 있는지, 아니면 그런 부분은 직접 생각하셔서 구현하신 건지 궁금합니다! (후자라면.. 이번 작업 이후로 진짜 네비게이션 완벽 마스터하신 거 아니예요?!?)

angryPodo replied:
하하 저도 NIA코드 보면서 왜쓰는거지? 했었는데요. 사용하는 이유를 보면 알 수 있다고 생각합니다.
NIA가 `@Composable` 속성을 사용한 이유 == '애니메이션 안정성' 인것 같더라구요.
밑에 부터는 제가 분석한 결과로 도출한 생각입니다ㅎㅎ

```kotlin
// NIA 코드
private val previousDestination = mutableStateOf<NavDestination?>(null)
 
 val currentDestination: NavDestination?
     @Composable get() {
         val currentEntry = navController.currentBackStackEntryFlow
             .collectAsState(initial = null)
 
         // Fallback to previousDestination if currentEntry is null
         return currentEntry.value?.destination.also { destination ->
             if (destination != null) {
                 previousDestination.value = destination
             }
         } ?: previousDestination.value // <--- 바로 이 부분!
     }
```

NIA는 `previousDestination`이라는 일종의 '캐시'를 사용하는데 이게 필요한 이유는 화면 전환 애니메이션(특히 Exit Animation)이 실행되는 동안의 안정성 때문입니다.

   1. 화면 A에서 화면 B로 이동할 때, 화면 A가 사라지는 애니메이션이 실행되는 아주 짧은 순간에
      `navController.currentBackStackEntry`는 일시적으로 `null`이 될 수 있습니다.
   2. 만약 이때 `currentDestination`에 의존하는 `UI`(예: 하단 탭바의 아이콘 하이라이트)가 있다면,
      `currentDestination`이 `null`이 되면서 하이라이트가 깜빡이거나 잠시 사라지는 현상이 발생할 수
      있습니다.
   3. NIA는 이 문제를 해결하기 위해 `previousDestination`을 사용합니다. `currentEntry`가 잠시 `null`이
      되더라도, 직전에 유효했던 `destination` 값을 `previousDestination`에 저장해두었다가 돌려주는
      것.
이 때문에 `@Composable`을 사용했다고 생각합니다. 그 외의 방법은 저도 안떠오르네요.

그래서 최대한 UI종속성을 제거하자 라고 생각해서 `stateIn`으로 바꿨습니다. 딱히 참고한 다른 레퍼런스는 모르겠어요ㅋㅋ..NIA만 참고했습니다. 어쨌든 제 목표는 `MainAppState`를 UI와 완전하게 분리된 테스터블한 `State Holder`로 만들고자 작업했습니다. 잘된건지는 모르겠네요ㅎㅎ

nahy-512 reviewed (approved):
진짜 최고.. 멘토링 때 현우님이 네비 코드 고쳤으면 좋겠다고 하신 거 듣고 어떻게 고치지 되게 막막했었는데.. 확실히 제가 고칠 수 있는 영역이 아니었던 것 같네요. 이렇게 빨리 수정하시다니ㅠㅠ 진짜 대단하십니다. 고생 너무 많으셨어요!

Hyobeen-Park reviewed (commented):
흠 근데 뷰가 많아질수록 MainAppState가 너무 커질 것 같기도 하고,,,, navigation 외에 다른 코드들이 많아질 일이 많이 없을 것 같아서 괜찮을 것 같기도 하고,,, 고민 좀 더 해보고 올게요..!!ㅎㅎㅎ

angryPodo committed:
[MOD/#182] 리뷰 반영

Hyobeen-Park reviewed (approved):
어푸푸🚀🚀

angryPodo committed:
[MOD/#182] Improve navigation state management in MainAppState

This commit refactors the navigation state management within `MainAppState.kt`.
